### PR TITLE
Add support for OP-TEE TA signature stitching during the build process.

### PR DESCRIPTION
**Pull request by aleksandarn-ee**
Signed-off-by: Aleksandar Nikolic <aleksandar.nikolic@mixed-mode.de>

The current build process does not allow stitching of the signature to the OP-TEE TA (e.g. trusted_keys), but only signing. The patch allows the user to sign the OP-TEE TAs (or more accurately the .stripped.elf files) offline and stitch the generated signature during the OP-TEE build process.

**Usage**

The following environment variables are to be defined by the user:

```
CFG_STITCH_TA=y     # Perform the signature stitching instead of TA signing.
TA_PUBLIC_KEY       # Path to the public key used for signature verification. Defaults to TA_SIGN_KEY.
TA_SIG_DIR          # Path to the directory where the signatures are. Defaults to ./signatures. The stitching process expects the signature files to be named as: <ta-uuid>.sig
TA_SIGN_ALG         # Hash and signing algorithm. Defaults to TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256 (same as in sign_encrypt.py).
```

**Example in Yocto**

```
EXTRA_OEMAKE = " \
    ...
    CFG_STITCH_TA=y \
    TA_PUBLIC_KEY=<path to the public key> \
    TA_SIG_DIR=<my-custom-sig-dir> \
    TA_SIGN_ALG=TEE_ALG_RSASSA_PKCS1_V1_5_SHA256 \
    ...
"
```

**Additional improvement**

With this patch the OPTEE build process expects the signature files to be named as: `<ta-uuid>.sig`. In my opinion this shouldn't be an issue for the user. However, if someone does not agree with this, I could adjust the patch so the user can give the list of signatures to be used during the stitching process (this makes things complicated imho)

### CODE REVIEW

File: ta/arch/arm/link.mk [hunk 1]
```diff
@@ -108,9 +116,15 @@ $(link-out-dir$(sm))/$(user-ta-uuid).ta: \
 			$(TA_SIGN_KEY) \
 			$(lastword $(SIGN_ENC))
 	@$(cmd-echo-silent) '  $$(cmd-echo$(user-ta-uuid)) $$@'
+ifeq ($(CFG_STITCH_TA),y)
+	$(q)$(SIGN_ENC) stitch --key $(TA_PUBLIC_KEY) $$(crypt-args$(user-ta-uuid)) \
+        --uuid $(user-ta-uuid) --sig $(TA_SIG_DIR)/$(user-ta-uuid).sig \
+		--algo $(TA_SIGN_ALG) --in $$< --out $$@
```
Comment (etienne-lms) [inline ta/arch/arm/link.mk:122]: Here the TA could be re-built but stitched with the signature from a likely previous _signatures/_ files build. You would need to recompute the signature once `$(user-ta-uuid).stripped.elf` is updated?

If you're considering using a key behind a pkcs11 token interface, `sing_encryp.py` could be updated to talk straight to the pkcs11 token for signing the TA(s). This would maybe remove the need for signature stitching.

Comment (aleksandarn-ee) [reply ta/arch/arm/link.mk:122]: > Here the TA could be re-built but stitched with the signature from a likely previous signatures/ files build. You would need to recompute the signature once $(user-ta-uuid).stripped.elf is updated?

Yes, if an OP-TEE TA changes, we have to recompute the signature. However, I don't think this is very important, since the OP-TEE TAs remain the same within a release. If I update the OP-TEE to another version, there is more work to do than just recomputing the signatures. Plus there is the same issue with user TAs, which change as much as often.

> If you're considering using a key behind a pkcs11 token interface, sing_encryp.py could be updated to talk straight to the pkcs11 token for signing the TA(s). This would maybe remove the need for signature stitching.

Good idea, I see minor issues with it though. 
* In our case I have no access to the HSM (its on a remote serve), so I wouldn't be able to build the OP-TEE this way, and the guy who does have access to it, has very little time to deal with it. 
* Do you think we could implement this in a generic way so everyone could sign the TAs without modifying the `sign_encrypt.py` script? A possible solution would be that the user should provide a module which would implement some specific stuff related to the pkcs11 token interface.  
* If we were to go this way, I'd have to see when I could implement it, since I have burned through the hours I had for this topic.

Comment (jenswi-linaro) [reply ta/arch/arm/link.mk:122]: > > Here the TA could be re-built but stitched with the signature from a likely previous signatures/ files build. You would need to recompute the signature once $(user-ta-uuid).stripped.elf is updated?
> 
> Yes, if an OP-TEE TA changes, we have to recompute the signature. However, I don't think this is very important, since the OP-TEE TAs remain the same within a release. If I update the OP-TEE to another version, there is more work to do than just recomputing the signatures. Plus there is the same issue with user TAs, which change as much as often.

What do you mean with "changes"? If you get a different binary when recompiling the TA? If you expect the same binary each time, why bother recompiling it?

Comment (aleksandarn-ee) [reply ta/arch/arm/link.mk:122]: > What do you mean with "changes"? If you get a different binary when recompiling the TA? If you expect the same binary each time, why bother recompiling it?

With "changes" I mean if the source code of a TA changes (a bug is fixed or new features are added) - happens only when a new version is released. In that case if we were to use an old signature, the stitching would fail. So we need to build the OP-TEE once, compute the hashes of the `<ta-uuid>.stripped.elf` files and sign them. Once done, the build process will always succeed, since the `<ta-uuid>.stripped.elf` files (and hence their hashes and signatures) do not change.

Comment (etienne-lms) [reply ta/arch/arm/link.mk:122]: It seems to me that you need to install pre-built/pre-signed TA binaries, rather than stitching some bin's for them to be installed.

Comment (aleksandarn-ee) [reply ta/arch/arm/link.mk:122]: > It seems to me that you need to install pre-built/pre-signed TA binaries, rather than stitching some bin's for them to be installed.

What do you mean to install pre-built/pre-signed TA binaries? The build process does not allow that, as it always builds and tries to sign the TAs. Or you mean to let the build fail on purpose and to sign the `.stripped.elf` files afterwards?

The reason for the PR is the following: the offline-signing procedure requires that the key `./keys/default_ta.pem` is a public key. If that is the case, the OPTEE build process fails with error: `ERROR:sign_encrypt.py:Provided key cannot be used for signing, please use offline-signing mode.`. The reason for that is that the build process tries to sign the TAs with the public key and fails. 

To sum up: OPTEE OS cannot be compiled if `./keys/default_ta.pem` is a public key, what is required for offline-signing.

**Comment by etienne-lms** [comment]
Ok. Maybe a a a new config switch CFG_TEE_SIGN_TAS, default `y` and if not then **ta.mk** build only generates the TA elf binaries.

> What do you mean to install pre-built/pre-signed TA binaries? 

I understood that you don't have access to the singing process at build time, so you use previously generated signatures (from path **.../signatures/**) and I belive such prebuilt signature can only be stitched to its previously built TA elf binary. 
So in the end, you install pre-signed TAs, no?
I

**Comment by aleksandarn-ee** [comment]
> I understood that you don't have access to the singing process at build time, so you use previously generated signatures (from path **.../signatures/**) and I belive such prebuilt signature can only be stitched to its previously built TA elf binary.

You understood correctly. 

> So in the end, you install pre-signed TAs, no?

In the end I want that OPTEE build succeeds and, in case of offline signing, produces the TA .ta binaries. For that reason I need the patch from above. (Take a look at the answer below).

> Ok. Maybe a a a new config switch `CFG_TEE_SIGN_TAS`, default `y` and if not then **ta.mk** build only generates the TA elf binaries **?**
> **(edited) add the question mark**
> 

This is, in my opinion, even a better solution, as the it might be a bit more straightforward for the user.  So in case CFG_TEE_SIGN_TAS is set to `n`, no signing takes place and the OPTEE build process builds only the TA elf files. It is up to the user to stitch the signatures to the TAs after OPTEE has been built (or to sign or whatever)?

**Comment by aleksandarn-ee** [comment]
@etienne-lms I pushed a new commit. Does it satisfy your requirements?

**Comment by etienne-lms** [comment]
> > If you're considering using a key behind a pkcs11 token interface, sing_encryp.py could be updated to talk straight to the pkcs11 token for signing the TA(s). This would maybe remove the need for signature stitching.
> 
> Good idea, I see minor issues with it though.
>
> In our case I have no access to the HSM (its on a remote serve), so I wouldn't be able to build the OP-TEE this way, and the guy who does have access to it, has very little time to deal with it.

In your case, I think you should use OP-TEE/build.git to get everything built **but** the TAs, and have means to embedded your prebuilt and pre-signed TAs from a private staging directory.


> Do you think we could implement this in a generic way so everyone could sign the TAs without modifying the sign_encrypt.py script?

Yes. Generic _sign_encypt.py_ could use pkcs11 API instead of module pycrypto_xxx_ + clear private key.
Of course this would be useful to environments where the pkcs11 token can be reached (which is not your case).

I have something for that that could help:
changes in _scripts/sign_encrypt.py_: https://github.com/etienne-lms/optee_os/commit/cc9199f7a76826cd81021e4a350d4052c84d6847
changes in  _ta/arch/arm/link.mk_: https://github.com/etienne-lms/optee_os/commit/c2429939d8211120ddf481e784232be482a48540 (introduces CFG_SIGN_TA_WITH_PKCS11)

These 2 changes allows one to build TAs and sign them with a pkcs11 token, using somehting like:
`PKCS11_MODULE=/path/to/<pkcs11_lib>.so PKCS11_TOKEN_LABEL="some-label" PKCS11_KEY_ID=<some-value> make all`


> A possible solution would be that the user should provide a module which would implement some specific stuff related to the pkcs11 token interface.
> If we were to go this way, I'd have to see when I could implement it, since I have burned through the hours I had for this topic.

To fake or emulate a pkcs11 token, you could use [Softhsm2](https://github.com/opendnssec/SoftHSMv2). But I don't think it's worth emulating a pkcs11 token if it can store to needed private key.

File: ta/arch/arm/link.mk (hunk 1 above)
Comment (etienne-lms) [reply ta/arch/arm/link.mk]: I agree that OP-TEE should be able to build without mandating signature at build time.
Building with no `TA_SIGN_KEY` as far as at least **TA_PUBLIC_KEY** is defined (to get the public key embedded in Core) should not fail.

**Comment by etienne-lms** [comment]
> @etienne-lms I pushed a new commit. Does it satisfy your requirements?

I'm not sure. This makes build to do generate `*.ta` files while these are still pushed in make target list. I'm not sure that is really nice. I would prefer @jforissier view on that.

Maybe you should call makefile targets that don't rebuild at all the TAs, not the `all` default target that builds everything.

**Comment by jforissier** [comment]
> > @etienne-lms I pushed a new commit. Does it satisfy your requirements?
> 
> I'm not sure. This makes build to do generate `*.ta` files while these are still pushed in make target list. I'm not sure that is really nice. I would prefer @jforissier view on that.
> 
> Maybe you should call makefile targets that don't rebuild at all the TAs, not the `all` default target that builds everything.

+1

Currently the meaningful targets are all, clean, cscope, ldelf, mem_usage and ta_dev_kit (you can see them via auto-completion by typing `make <TAB><TAB>`). ta_dev_kit contains the user space include files and libraries for TA development, as well as the in-tree TAs.
We could very well create intermediate targets such as `ta` and `teecore` if that helps.

**Comment by aleksandarn-ee** [comment]
> Currently the meaningful targets are all, clean, cscope, ldelf, mem_usage and ta_dev_kit (you can see them via auto-completion by typing `make <TAB><TAB>`). ta_dev_kit contains the user space include files and libraries for TA development, as well as the in-tree TAs.
> We could very well create intermediate targets such as `ta` and `teecore` if that helps.

Okay, great. What would the `ta` and `teecore` targets do? I guess `teecore` would only build the OPTEE binaries, but `ta`? And what would be the difference between `ta` and `ta_dev_kit`?

**Comment by jforissier** [comment]
> And what would be the difference between `ta` and `ta_dev_kit`?

It is just an example, maybe not very good because currently `ta_dev_kit` includes the *.ta files.

The point is, I find it preferable to introduce a new target rather than a new config variable when it comes to telling the build system what to build (as opposed to how to build it). Please note that adding a `CFG_` to the build command will cause recompilation of a lot of things. So, `make all_but_signed_tas` instead of `make [all] CFG_TEE_SIGN_TAS=n` would be better.

**Comment by aleksandarn-ee** [comment]
> > And what would be the difference between `ta` and `ta_dev_kit`?
> 
> It is just an example, maybe not very good because currently `ta_dev_kit` includes the *.ta files.
> 
> The point is, I find it preferable to introduce a new target rather than a new config variable when it comes to telling the build system what to build (as opposed to how to build it). Please note that adding a `CFG_` to the build command will cause recompilation of a lot of things. So, `make all_but_signed_tas` instead of `make [all] CFG_TEE_SIGN_TAS=n` would be better.

Understood. I'll push a new commit today or tomorrow and we can discuss this further.

**Comment by aleksandarn-ee** [comment]
> > And what would be the difference between `ta` and `ta_dev_kit`?
> 
> It is just an example, maybe not very good because currently `ta_dev_kit` includes the *.ta files.
> 
> The point is, I find it preferable to introduce a new target rather than a new config variable when it comes to telling the build system what to build (as opposed to how to build it). Please note that adding a `CFG_` to the build command will cause recompilation of a lot of things. So, `make all_but_signed_tas` instead of `make [all] CFG_TEE_SIGN_TAS=n` would be better.

@jforissier I gave a look at the compile process of OPTEE and, as I am no expert in Makefile, I don't know how to add another target. Would it be something you could implement?

Also, what do you think about my first [commit](https://github.com/OP-TEE/optee_os/pull/4614/commits/bbe58e80309042967ce2ea7c0ecdfa19e2e00fe0)? Via the environment variables the user can define **how** the .ta files are built (either by signing or by stitching), in both cases however `make all` would build the .ta files.

File: ta/arch/arm/link.mk [hunk 2]
```diff
@@ -5,6 +5,14 @@ link-script-dep$(sm) = $(link-out-dir$(sm))/.ta.ld.d
 SIGN_ENC ?= $(PYTHON3) $(ta-dev-kit-dir$(sm))/scripts/sign_encrypt.py
 TA_SIGN_KEY ?= $(ta-dev-kit-dir$(sm))/keys/default_ta.pem
 
+# In case the TA offline-signing has been performed.
+ifeq ($(CFG_STITCH_TA),y)
```
Comment (jforissier) [inline ta/arch/arm/link.mk]: You shouldn't use a CFG_ prefix due to the recompilation issue it triggers. Same reason we use `EARLY_TA_PATHS` rather than `CFG_EARLY_TA_PATHS` for instance. These variables are intended for the build system and should have no side-effect on the compiled source code.
I realize that `CFG_ENCRYPT_TA` has the same issue :-/

File: ta/arch/arm/link.mk [hunk 3]
```diff
@@ -5,6 +5,14 @@ link-script-dep$(sm) = $(link-out-dir$(sm))/.ta.ld.d
...
+# The signatures directory. The linking process expects <ta-uuid>.sig files in it.
+TA_SIG_DIR ?= signatures
+# The hash and signing algorithm.
+TA_SIGN_ALG ?= TEE_ALG_RSASSA_PKCS1_PSS_MGF1_SHA256
```
Comment (jforissier) [inline ta/arch/arm/link.mk]: I would prefer
```make
ifneq (,$(TA_SIGN_ALG))
crypt-args$(user-ta-uuid) += --algo $(TA_SIGN_ALG)
endif
```
...so that the default algorithm stays in `sign_encrypt.py` and is not duplicated here.

**Review by jforissier — COMMENTED** [review]
Would the following work for you? I am proposing to introduce two new targets: `ta_elf` and `ta`, as well as a `TA_SIGN_METHOD` variable (undefined means default, `stitch` for signature stitching).
```
# 1. Build the *.stripped.elf files
make ta_elf
# 2. Offline signing
run_some_external_script *.ta.elf -o /path/to/signatures
# 3. Generate *.ta using the 'stitch' method of sign_encrypt.py
# Remove the 'ta' target to build everything (TA dev kit etc.), 'all' depends on 'ta'
make [ta] TA_SIGN_METHOD=stitch [TA_SIG_DIR=/path/to/signatures] [TA_SIGN_ALG=xyz]
```

File: ta/arch/arm/link.mk (hunk 2 above)
Comment (aleksandarn-ee) [reply ta/arch/arm/link.mk]: Yep, the commit was before you mentioned the purpose of the CFG_XXX variables :)

File: ta/arch/arm/link.mk (hunk 3 above)
Comment (aleksandarn-ee) [reply ta/arch/arm/link.mk]: Makes sense of cource, much better!

**Comment by aleksandarn-ee** [comment]
> Would the following work for you? I am proposing to introduce two new targets: `ta_elf` and `ta`, as well as a `TA_SIGN_METHOD` variable (undefined means default, `stitch` for signature stitching).
> 
> ```
> # 1. Build the *.stripped.elf files
> make ta_elf
> # 2. Offline signing
> run_some_external_script *.ta.elf -o /path/to/signatures
> # 3. Generate *.ta using the 'stitch' method of sign_encrypt.py
> # Remove the 'ta' target to build everything (TA dev kit etc.), 'all' depends on 'ta'
> make [ta] TA_SIGN_METHOD=stitch [TA_SIG_DIR=/path/to/signatures] [TA_SIGN_ALG=xyz]
> ```

This would do the job. Could you implement the `ta_elf` and `ta` targets, and I will update the `ta/arch/arm/link.mk` as you proposed above and test the building with the new targets? I could also provide an update for the OPTEE documentation if necessary.

**Comment by jforissier** [comment]
@aleksandarn-ee cool, sounds good. I will propose something when I have a few spare cycles.

**Comment by aleksandarn-ee** [comment]
@jforissier Great. I will wait on your commit. Additionally, could you estimate when you will have a few spare cycles, so I could plan on my side? E.g. in a week, month, a few months, etc.

**Comment by jforissier** [comment]
@aleksandarn-ee probably tomorrow or Friday.

**Comment by aleksandarn-ee** [comment]
@jforissier Great :)

File: ta/arch/arm/link.mk (hunk 3 above)
Comment (jforissier) [reply ta/arch/arm/link.mk]: And we could (should) even wait for a real use case before adding this variable. Do you plan on using a different signing scheme?

**Comment by jforissier** [comment]
@aleksandarn-ee please see/test/comment https://github.com/OP-TEE/optee_os/pull/4632. Thanks!

File: ta/arch/arm/link.mk (hunk 3 above)
Comment (aleksandarn-ee) [reply ta/arch/arm/link.mk]: Yes, our HSM server uses (at the moment) the PKCS1.5, so without this variable we wouldn't be able to stitch the signature. Id add this variable and remove it if/when the `sign_encrypt.py` script supports exactly one signature scheme. If however you think it is not necessary, I can add it locally just in my case.

**Comment by aleksandarn-ee** [comment]
@jforissier this is not working for us. The problem is compiling the OPTEE core. To my knowledge compiling the OPTEE core is possible only with `make all` (correct me if I am wrong). In order to have the correct public key embedded in the OPTEE core we need that `./keys/ta_default.pm` is a custom public key (also used for stitching later). When that is the case, `make all` will fail as the `all` target tries to sign the `.stripped.elf` files by default.

Suggestion:
* `make all_but_signed_ta` (or some other name for the target) - compiles the OPTEE core and TAs up to the `.stripped.elf` files
* `make ta` - remains as you proposed. It either signs or stitches.

When the user does not need the stitching he can just call `make all`.

**Comment by jforissier** [comment]
> The problem is compiling the OPTEE core. To my knowledge compiling the OPTEE core is possible only with `make all` (correct me if I am wrong)

Not correct. It is a makefile so you can ask to build whatever file you need. For example:
```
make out/arm-plat-vexpress/core/tee-header_v2.bin out/arm-plat-vexpress/core/tee-pager_v2.bin out/arm-plat-vexpress/core/tee-pageable_v2.bin
```
...which can be shortened like so:
```
make out/arm-plat-vexpress/core/tee-{header,pager,pageable}_v2.bin
```
(Sidenote: shell completion works with makefiles too, which I find pretty awesome. Try `rm -rf out` then `make out/arm-plat-vexpress/core/tee<TAB><TAB>` :smiley:)

We could add a `tee_bin` target as a synonym to the above line if that helps?

**Comment by aleksandarn-ee** [comment]
> > The problem is compiling the OPTEE core. To my knowledge compiling the OPTEE core is possible only with `make all` (correct me if I am wrong)
> 
> Not correct. It is a makefile so you can ask to build whatever file you need. For example:
> 
> ```
> make out/arm-plat-vexpress/core/tee-header_v2.bin out/arm-plat-vexpress/core/tee-pager_v2.bin out/arm-plat-vexpress/core/tee-pageable_v2.bin
> ```
> 
> ...which can be shortened like so:
> 
> ```
> make out/arm-plat-vexpress/core/tee-{header,pager,pageable}_v2.bin
> ```
> 
> (Sidenote: shell completion works with makefiles too, which I find pretty awesome. Try `rm -rf out` then `make out/arm-plat-vexpress/core/tee<TAB><TAB>` smiley)
> 
Thanks for the clarification!

> We could add a `tee_bin` target as a synonym to the above line if that helps?

That'd be great.

**Comment by aleksandarn-ee** [comment]
Closed in favor of this [PR](https://github.com/OP-TEE/optee_os/pull/4632).
